python/single_cell_model.cpp

```diff
@@ -0,0 +1,230 @@
+
+    // example use:
+    //      m.probe('voltage', arbor.location(2,0.5))
+    void probe(const std::string& what, const arb::mlocation& where, double frequency) {
+        if (what != "voltage") {
+            throw pyarb_error(
+                util::pprintf("{} does not name a valid variable to trace (currently only 'voltage' is supported)", what));
+        }
+        if (frequency<=0) {
+            throw pyarb_error(
+                util::pprintf("sampling frequency is not greater than zero", what));
+        }
+        if (where.branch>=cell_.morphology().num_branches()) {
+            throw pyarb_error(
+                util::pprintf("invalid location", what));
+        }
+        probes_.push_back({where, frequency});
+    }
+
+    void run(double tfinal) {
+
+        single_cell_recipe rec(cell_, probes_, generators_);
+
+        auto domdec = arb::partition_load_balance(rec, ctx_);
+
+        sim_ = std::make_unique<arb::simulation>(rec, domdec, ctx_);
+
+        // Create one trace for each probe.
+        traces_.reserve(probes_.size());
+
+        // Add probes
+        for (arb::cell_lid_type i=0; i<probes_.size(); ++i) {
+            const auto& p = probes_[i];
+
+            traces_.push_back({"voltage", p.site, {}, {}});
+
+            auto sched = arb::regular_schedule(1000./p.frequency);
+
+            // Now attach the sampler at probe site, with sampling schedule sched, writing to voltage
+            sim_->add_sampler(arb::one_probe({0,i}), sched, trace_callback(traces_[i]));
+        }
+
+        // Set callback that records spike times.
+        sim_->set_global_spike_callback(
+            [this](const std::vector<arb::spike>& spikes) {
+                for (auto& s: spikes) {
+                    spike_times_.push_back(s.time);
+                }
+            });
+
+        sim_->run(tfinal, 0.025);
+
+        run_ = true;
+    }
+
+    const std::vector<double>& spike_times() const {
+        return spike_times_;
+    }
+
+    const std::vector<trace>& traces() const {
+        return traces_;
+    }
+};
+
+void register_single_cell(pybind11::module& m) {
+    using namespace pybind11::literals;
+
+    pybind11::class_<trace> tr(m, "trace");
+    tr
+        .def_readonly("variable", &trace::variable)
+        .def_readonly("location", &trace::loc)
+        .def_readonly("time",    &trace::t)
+        .def_readonly("value",   &trace::v);
+
+    pybind11::class_<single_cell_model> model(m, "single_cell_model",
+        "Wrapper for easily building and running single cell models.");
+
+    model
+        .def(pybind11::init<arb::cable_cell>(),
+            "cell"_a, "Build a single cell model.")
+        .def("run", &single_cell_model::run, "tfinal"_a, "run model from t=0 to t=tfinal ms")
+        .def("probe", &single_cell_model::probe, "what"_a, "where"_a, "frequency"_a)
+        .def_property_readonly("spikes", [](const single_cell_model& m) {return m.spike_times();}, "spike times (ms)")
```

Comment:
capital letters, build sentences, use [] for units e.g.`Return spike times [ms].`